cocos/network/WebSocketImpl.cpp

```diff
@@ -0,0 +1,940 @@
+*/
+#define CHECK_INVOKE_FLAG(flg)  do { \
+    if(_callbackInvokeFlags & flg)  return 0; \
+    _callbackInvokeFlags |= flg; } while(0)
+
+#define LOCK_MTX(mtx) std::lock_guard<std::mutex> __guard__(mtx)
+
+//////////////////////////tool functions ////////////////////////
+
+//android only
+static std::string fileBaseName(const std::string &path)
+{
+    auto lastSlashIdx = path.find_last_of("\\/");
+    if (lastSlashIdx == std::string::npos) return path;
+    return path.substr(lastSlashIdx + 1);
+}
+/**
+* test if path starts with "https" or "wss"
+*/
+static bool isSecureURL(const std::string &path)
+{
+    const char *p = path.c_str();
+    if (path.length() < 5) return false;
+
+    switch (p[0])
+    {
+    case 'h':
+    case 'H':
+    case 'w':
+    case 'W':
+    {
+        if (strncmp(p, "https:", 6) == 0 ||
+            strncmp(p, "wss:", 4) == 0 ||
+            strncmp(p, "HTTPS:", 6) == 0 ||
+            strncmp(p, "WSS:", 4) == 0
+            ) {
+            return true;
+        }
+        break;
+    }
+    default:
+        break;
+    }
+    return false;
+}
+
+namespace cocos2d
+{
+    namespace network
+    {
+        /**
+        * used to present whether `WebSocketImpl#_delegate` callback ever be called.
+        */
+        enum CallbackInvoke {
+            CallbackInvoke_CONNECTED = 1 << 0,
+            CallbackInvoke_CLOSED = 1 << 1,
+            CallbackInvoke_ERROR = 1 << 2
+        };
+
+
+        ////////////////////net thread - begin ///////////////////
+
+        //////////////basic data type - begin /////////////
+        enum class NetCmdType
+        {
+            OPEN, CLOSE, WRITE, RECIEVE
+        };
+
+        class NetDataPack {
+        public:
+            NetDataPack() {}
+            NetDataPack(const char *f, size_t l, bool isBinary) {
+                _buffer = (uint8_t*)calloc(1, l + LWS_PRE);
+                memcpy(_buffer + LWS_PRE, f, l);
+                _size = l;
+                _remain = l;
+                _payload = _buffer + LWS_PRE;
+                _isBinary = isBinary;
+            }
+
+            NetDataPack(const NetDataPack &) = delete; //non-copyable
+            NetDataPack(NetDataPack&&) = delete;       //non-movable
+
+            ~NetDataPack() {
+                if (_buffer) {
+                    free(_buffer);
+                    _buffer = nullptr;
+                }
+                _size = 0;
+            }
+
+
+            size_t remain() { return _remain; }
+            uint8_t *payload() { return _payload; }
+
+            void consume(size_t d)
+            {
+                CCASSERT(d <= _remain, "data not enought!");
+                _payload += d;
+                _remain -= d;
+                _consumed += d;
+            }
+
+            size_t consumed() { return _consumed; }
+            bool isBinary() { return _isBinary; }
+        private:
+            uint8_t *_buffer = nullptr;
```

Comment:
why not std::vector<uint8_t> buffer; ?

Reply:
`_buffer` is a plain linear buffer, no resize operation during the lifetime. I think the pointer is simpler.